wgengine/magicsock/magicsock.go

```diff
@@ -1360,6 +1387,17 @@ Top:
 		c.mu.Lock()
 		if dk, ok := c.discoOfNode[tailcfg.NodeKey(dm.src)]; ok {
 			discoEp = c.endpointOfDisco[dk]
+			// If we know about the node (it's in discoOfNode) but don't know about the
+			// endpoint, that's because it's an idle peer that doesn't yet exist in the
+			// wireguard config. So run the receive hook, if defined, which should
+			// create the wireguard peer.
+			if discoEp == nil && c.noteRecvActivity != nil {
+				c.mu.Unlock()          // release lock before calling noteRecvActivity
+				c.noteRecvActivity(dk) // (calls back into CreateEndpoint)
+				c.mu.Lock()            // reacquire lock
```

Comment:
What else might have raced in and mutated magicsock state During this unlocked period?

Unlocks mid-function like this make me nervous unless it's explicitly doing some kind of condvar wait thing, because the struct is effectively in a completely knew and unknown state once we relock. Based on comments earlier up, I assume you worked through alternatives and decided this was the least bad, but can you add some comments warning future readers to remember that the state after this reacquire may be entirely unrelated to that from before the acquire?

Reply:
Done.